### ISSUE TITLE
how to choose imgsz

### ISSUE THREAD

**Issue by lucas-mior** [ultralytics/ultralytics#1037]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Assuming that I should train the model with images in two different resolutions:
640x400
400x640

How should I specify `imgsz` (both in training and prediction) in order to maximize prediction accuracy?
Say I will have 5000 images in each resolution.

**Comment by FrancescoSaverioZuppichini** [comment]
Hi 👋 

Reading the [doc](https://docs.ultralytics.com/cli/) is usually a good place to start before asking :) But, we are engineers and we don't like docs hahaha

you can use `imgsz` params, 

```
yolo train ... imgsz 640
```

`640x640` is a good resolution, stick with it and call it a day

**Comment by lucas-mior** [comment]
@FrancescoSaverioZuppichini thanks for the answer.
I don't think you understand what I asked. I know how to pass arguments to yolo (using `imgsz=640`).
What I'm asking is *which* image size would be more apropriate in my case.

You are suggesting to use `imgsz=640,640` so that all my input images will be distorted to fit in that shape?
Thanks in advance.

For reference:
![imgsz](https://user-images.githubusercontent.com/49657055/220427078-7adcd1ed-bf6d-467c-af3f-18271ad87f6e.jpg)

**Comment by FrancescoSaverioZuppichini** [comment]
@lucas-mior thanks for the highlights :) Sorry for the misunderstanding. I'd say you keep  `640x640` for training, since yolo (if I am not mistaken) was trained on that shape specifically (I mean, random crops of that shape I think). 

By looking at the code, it doesn't seems possible to pass a tuple of shapes (e.g. 640x400). I am not 100% sure how the size is used internally, I do believe your images will be mixed together and the final shape will be `640x640` so no big distortion should happen.

I believe we can safely assert it by running a dummy train and check for the resulting batch images stores somewhere in the result folder. Does it make sense to you?

Curious to actually see what is going on under the hood 😄

**Comment by AyushExel** [comment]
Agree with @FrancescoSaverioZuppichini that you should use 640 imgsz by default unless needed otherwise. Increasing the training resolution might increase accuracy but it slows the training too much and in most cases, the tradeoff might not be worth it. For instance, the p6 models trained at 1280 imgsz will take more than double the time taken by 640 models. Of course, you can use it if more time/hardware doesn't matter. 

Regarding your question about training at custom resolutions (640x400 in this case), you can enable rectangular training by passing `rect=True`. Training will look at your highest dimensions and it'll work out the other dimension by adjusting it for stride 32.
Oh, and even if you use 640x640 (default) your images should not be distorted by the interpolation operation. If you're seeing that in your visualization please raise a report and we'll look into it.

P.S. I'll be quite inactive till Friday so I'm tagging @Laughing-q in case there are any other questions about data loaders.

**Comment by Laughing-q** [comment]
> You are suggesting to use imgsz=640,640 so that all my input images will be distorted to fit in that shape?

@lucas-mior hi, since your images are `640x400` and `400x640`, I'd like to say just set `imgsz=640` for the first try, and your training images will automatically be resized and padded to `640x640` while training.
Also like @AyushExel said , your images won't be distorted, training images will be resized while keeping the aspect ratio and padded the rest part to get `640x640`.

**Comment by lucas-mior** [comment]
Thanks everyone for the information. 
Now I understand how imgsz works and can properly choose for my use case.

**Comment by pderrenger** [comment]
@lucas-mior you're welcome! If you have any more questions or need further assistance, feel free to ask. Good luck with your training! 🚀

**Comment by vonzo** [comment]
What about smallezr images? 

I'm trainning images of around 200x200 with different ratios.

**Comment by glenn-jocher** [comment]
@vonzo when training and predicting with YOLOv8, the `imgsz` parameter defines the input image size that the model will expect. Choosing the right `imgsz` is important for maximizing prediction accuracy, as it can affect the model's ability to detect objects of various sizes within the images.

Here are some considerations for choosing `imgsz`:

1. **Aspect Ratio**: If your dataset contains images with different aspect ratios, you should choose an `imgsz` that maintains the aspect ratio of the majority of your images or is close to the average aspect ratio. For your case, since you have images with aspect ratios of 640x400 and 400x640, you might consider using a square `imgsz` like 640x640 to accommodate both without significant distortion.

2. **Resolution**: Higher resolutions can capture more detail, which can be beneficial for detecting small objects. However, higher resolutions also require more computational resources. You should balance between the resolution that captures sufficient detail and the available computational power.

3. **Consistency**: It's generally a good practice to train and predict with the same `imgsz` to maintain consistency. If you train with one image size and predict with another, the model might not perform optimally.

4. **Memory Constraints**: Larger `imgsz` values will consume more memory. Ensure that your hardware can handle the chosen size without running out of memory during training or inference.

5. **Batch Size**: Larger image sizes may require you to reduce the batch size to fit the data into your GPU memory. This can affect the training dynamics, so you may need to adjust the learning rate accordingly.

Given your scenario with two different resolutions, you could consider the following approaches:

- **Square Size**: Use a square `imgsz` such as 640x640 for both training and prediction. This will involve resizing the images, which may introduce some distortion, but it will allow the model to train on a consistent image size.

- **Multiple Sizes**: YOLOv8 supports multi-size training, where `imgsz` can be set as a range during training (e.g., `imgsz=[640, 640]`), and the model will randomly resize images to different sizes within this range for each batch. This can help the model generalize better to different image sizes.

- **Custom Size**: If you want to avoid the square resizing, you could choose a custom size that is a compromise between the two aspect ratios, such as 640x512. This would minimize the resizing distortion for both aspect ratios.

In summary, for your case with equal numbers of two different resolutions, a square `imgsz` like 640x640 might be the simplest and most effective choice for both training and prediction. However, if computational resources allow, experimenting with multi-size training or a custom compromise size could potentially yield better results. Always validate the model's performance on a representative validation set to ensure that the chosen `imgsz` is effective.
